### ISSUE TITLE
SIM102 walrus operator

### ISSUE THREAD

**Issue by Kludex** [charliermarsh/ruff#2021]
# Description

SIM102 is being raised when using walrus operator.

## Code

```python
def something(schema: dict) -> None:
    if all_of := schema.get("allOf"):
        if len(all_of) == 1:
            schema.update(all_of[0])
            del schema["allOf"]


something({"allOf": [{"type": "string"}]})
something({})
```

## Output

```bash
main.py:2:5: SIM102 Use a single `if` statement instead of nested `if` statements
```

## Notes

An alternative code for the above would be:
```python
def something(schema: dict) -> None:
    all_of = schema.get("allOf")
    if all_of and len(all_of) == 1:
        schema.update(all_of[0])
        del schema["allOf"]


something({"allOf": [{"type": "string"}]})
something({})
```
Which passes the linter.

**Comment by charliermarsh** [comment]
Thanks for filing!

I think this actually does produce valid code -- running `ruff --isolated --select SIM102 --fix` on that snippet gives me:

```py
def something(schema: dict) -> None:
    if (all_of := schema.get("allOf")) and len(all_of) == 1:
        schema.update(all_of[0])
        del schema["allOf"]


something({"allOf": [{"type": "string"}]})
something({})
```

Which seems to run ok? As long as the walrus is parenthesized. Did you run into another example where this fix produced invalid code? (Or is the issue to suggest that, stylistically, this is worse than before?)

**Comment by Kludex** [comment]
Ah! This is fine. Thanks 🙏
